README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
refactor
========

[![Project Status: WIP ? Initial development is in progress, but there has not yet been a stable, usable release suitable for the public.](http://www.repostatus.org/badges/latest/wip.svg)](http://www.repostatus.org/#wip) [![Build Status](https://travis-ci.org/jonmcalder/refactor.svg?branch=master)](https://travis-ci.org/jonmcalder/refactor) [![codecov](https://codecov.io/gh/jonmcalder/refactor/branch/master/graph/badge.svg)](https://codecov.io/gh/jonmcalder/refactor)

------------------------------------------------------------------------

[![minimal R version](https://img.shields.io/badge/R%3E%3D-3.2.1-6666ff.svg)](https://cran.r-project.org/) [![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/refactor)](https://cran.r-project.org/package=refactor) [![packageversion](https://img.shields.io/badge/Package%20version-1.1.0-orange.svg?style=flat-square)](commits/master)

------------------------------------------------------------------------

[![Last-changedate](https://img.shields.io/badge/last%20change-2017--02--09-yellowgreen.svg)](/commits/master)

**refactor** is an R package which aims to provide better handling of factors. Though R does have a special data type for factors, it isn't always explicity catered for in commonly used R functions, which can lead to unexpected and undesirable outcomes (e.g. see this [Win-Vector blog post](http://www.win-vector.com/blog/2014/09/factors-are-not-first-class-citizens-in-r/)). This observation formed the inspiration for 're'-factor: which is essentially to 're'-visit functions likely to be used with factor data and where possible to wrap, extend or override them in order to better cater for factor data, or at the very least to provide warnings when the integrity of the data could be compromised by an operation.

Installation
------------

refactor is not yet available on CRAN but you can easily install the latest development version from github using `devtools`:

``` r
# install.packages("devtools")
devtools::install_github("jonmcalder/refactor")
```

Overview
--------

Below are a few examples to illustrate some scenarios in which refactor can improve on base R's handling of factors.

Please see the vignette for more details: `vignette("refactor", package = "refactor")`

#### Better factor creation with `cfactor()`

-   get warnings whenever provided factor levels don't fully match those present in the data

``` r
string <- c("a", "b", "c")
factor(string, levels = c("b", "c", "d"))
#> [1] <NA> b    c   
#> Levels: b c d
cfactor(string, levels = c("b", "c", "d"))
#> Warning: the following levels were empty: 
#>  d
#> Warning: the following levels were removed: 
#>  a
#> [1] <NA> b    c   
#> Levels: b c d
```

-   numerical sequences in strings are detected and utilized to obtain correct orderings for factor levels

``` r
hard_to_detect <- c("EUR 21 - EUR 22", "EUR 100 - 101", "EUR 1 - EUR 10", "EUR 11 - EUR 20")
factor(hard_to_detect, ordered = TRUE)
#> [1] EUR 21 - EUR 22 EUR 100 - 101   EUR 1 - EUR 10  EUR 11 - EUR 20
#> 4 Levels: EUR 1 - EUR 10 < EUR 100 - 101 < ... < EUR 21 - EUR 22
cfactor(hard_to_detect, ordered = TRUE)
#> [1] EUR 21 - EUR 22 EUR 100 - 101   EUR 1 - EUR 10  EUR 11 - EUR 20
#> 4 Levels: EUR 1 - EUR 10 < EUR 11 - EUR 20 < ... < EUR 100 - 101
```

#### Extension of the generic `cut` method to handle (discrete) integer data

-   generate 'natural' intervals for factors when using cut on integer vectors
    -   e.g. `[1,3], [4,6], [7,9]` as opposed to `(0.5, 3.5], (3.5, 6.5], (6.5, 9.5]`
-   label factor levels more intuitively
    -   e.g. `1-3, 4-6, 7-9` as opposed to `[1,3], [4,6], [7,9], or (0,3], (3,6], (6-9]`

``` r
x_int <- 1:9
cut.default(x_int, breaks = 3)
#> [1] (0.992,3.67] (0.992,3.67] (0.992,3.67] (3.67,6.33]  (3.67,6.33] 
#> [6] (3.67,6.33]  (6.33,9.01]  (6.33,9.01]  (6.33,9.01] 
#> Levels: (0.992,3.67] (3.67,6.33] (6.33,9.01]
cut(x_int, breaks = 3)
#> [1] 1-3 1-3 1-3 4-6 4-6 4-6 7-9 7-9 7-9
#> Levels: 1-3 4-6 7-9
```

#### Extension of the generic `cut` method to handle (non-numeric) ordered data

-   `cut` usually requires numeric data, but refactor extends `cut` to handle other ordered data

``` r
some_letters <- factor(c('d','e','f','a','b','c','g','h','i'), ordered = TRUE)
cut(some_letters, breaks = c('a','c','f','i'), include.lowest = TRUE, ordered_result = TRUE)
#> [1] d-f d-f d-f a-c a-c a-c g-i g-i g-i
#> Levels: a-c < d-f < g-i
```

Contributions
-------------

Suggestions and feedback are most welcome.

Feel free to [open an issue](https://github.com/jonmcalder/refactor/issues) if you want to request a feature or report a bug, and make a pull request if you can contribute.


NEWS.md
# refactor 0.1.1

* Update breaks_mode with new 'default' method to emulate cut.default
* Rename existing breaks_mode method to 'spread'
* Updates to documentation to reflect these functional changes and further 
  clarify behavior
* Updates to testing of cut.integer based on the new modes
* Updates to package vignette

### Bug Fixes

# refactor 0.1.0.9001

* Style: All lines of code are now (roughly) 80 characters long, T and F changed 
  into explicit notation TRUE and FALSE
  
#### cut.integer

* floor() is used rather than round() when breaks are supplied as decimal 
  numbers
* The quantile method has been dropped

#### index_cfactor

* Arguments are only recycled if length = 1 (previously full recycling was 
  possible)
* Improve formatting of documentation and corrected spelling 
* Add more details on how this packages methods deviate from base methods
* Eliminate undocumented function arguments
* More meaningful helpfile titles
* Wrap relevant expressions in documentation so they appear as code
* Remove remaining quantile method and descriptions snippets
* Update examples so they are in line with the vignette

# refactor 0.1.0.9000

* Substantial updates to cfactor, cut.integer and cut.ordered
* More assertive tests and warnings
* Add new function index_cfactor()
* Updates to function documentation
* Updates to README
* Include package vignette with examples of the problems addressed

# refactor 0.1.0

* Updates to cut.integer and improved documentation
* Additional testing
* Include checkmate package to facilitate runtime checking of function 
  arguments

vignettes/refactor.R
## ------------------------------------------------------------------------
lower <- factor(letters[1:3], levels = letters[2:4])
upper <- factor(LETTERS[1:5], levels = LETTERS[1:4])
(combined <- c(lower, upper))

## ------------------------------------------------------------------------
library(refactor)

string <- c("a", "b", "c")
cfactor(string, levels = c("b", "c", "d"))

## ------------------------------------------------------------------------
easy_to_dectect <- c("EUR 11 - EUR 20", "EUR 1 - EUR 10", "EUR 21 - EUR 22")
factor(easy_to_dectect, ordered = TRUE) # correctly detects level

## ------------------------------------------------------------------------
hard_to_dectect <- c("EUR 21 - EUR 22", "EUR 100 - 101", 
                     "EUR 1 - EUR 10", "EUR 11 - EUR 20")

factor(hard_to_dectect, ordered = TRUE)

## ------------------------------------------------------------------------
cfactor(hard_to_dectect, ordered = TRUE, sep = "-")

## ------------------------------------------------------------------------
identical(
  cfactor(hard_to_dectect, ordered = TRUE, sep = NULL),
   factor(hard_to_dectect, ordered = TRUE)
)

## ------------------------------------------------------------------------
cfactor(x = c("a", "b", "c"), levels = c("a", "b", "c"), labels = c("a", "letter b", "b"))

## ------------------------------------------------------------------------
data <- sample(x = 1:10, size = 20, replace = TRUE)
index <- data.frame(encoding = 1:10,
                    label = letters[1:10])

cfactor(data, levels = index$encoding, labels = index$label)

## ------------------------------------------------------------------------
data <- data.frame(var1 = sample(x = 1:10, size = 20, replace = TRUE),
                  var2 = rep(1:2, 20),
                  var3 = sample(20),
                  var4 = 2, 
                  var5 = sample(row.names(USArrests), size = 20),
                  stringsAsFactors = FALSE)
head(data)

index <- data.frame(var = rep(paste0("var", 1:3), c(10, 2, 20)),
                    encoding = c(1:10, 1:2, 1:20),
                    label = c(letters[1:10], c("male", "female"), LETTERS[1:20]))
head(index)

## ------------------------------------------------------------------------
final <- head(index_cfactor(data = data, index = index, variable = "var", 
                            ordered = c(TRUE, TRUE, FALSE)))

print(final)
sapply(final, class)

## ---- echo = FALSE, include = TRUE---------------------------------------
c("1 to 4", "5 to 6") # properly separated
c("1 to 4", "4 to 6") # not properly separated
c("from 1,000 to 2,000", "from 2000 to 4,000") # comma separated and 'from' and 'to'
c("4.0 / 4.1", "4.2 / 4.3") # point and slash separator
c("one minute", "three minutes", "1 hour")

## ------------------------------------------------------------------------
random <- sample(100)
cut.default(random, breaks = seq(0, 100, by = 10))[1:10]

## ------------------------------------------------------------------------
cut(random, breaks = seq(0, 100, by = 10))[1:10]

## ------------------------------------------------------------------------
cut(sample(10), breaks = c(0, 3, 5))

## ------------------------------------------------------------------------
cut(sample(10), breaks = c(1, 4, 6, 8, 9, 10))

## ------------------------------------------------------------------------
cut(sample(10), breaks = c(10, 0, 3))

## ------------------------------------------------------------------------
cut(sample(10), breaks = c(1, 2.6, 5.1, 10))

## ------------------------------------------------------------------------
some_letters <- cfactor(letters, ordered = TRUE)
head(cut(some_letters, breaks = c("a", "q", "z"), 
         labels = c("beginning of the alphabet", "the rest of the alphabeth"), 
         right = TRUE, include.lowest = TRUE))

